Integrate GIS shape file creation

yerbol-akhmetov opened the pull request:
Closes #2 

## Changes proposed in this Pull Request
Good day. This PR aims to integrate GIS shape file creation scripts to the repository. The rules prepare country shape of the UK and splits the shapefile into regions based on the ETYS boundaries. Then, additional manual splitting of regions are accomplished based on latitudes and longitudes with a final merge.

## Checklist

- [x] I tested my contribution locally and it works as intended.
- [ ] Code and workflow changes are sufficiently documented.
- [x] Changed dependencies are added to `envs/environment.yaml`.
- [ ] Changes in configuration options are added in `config/config.default.yaml`.
- [ ] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [x] OET license identifier is added to all edited or newly created code files.
- [ ] Sources of newly added data are documented in `doc/data_sources.rst`.
- [x] A release note `doc/release_notes.rst` is added.

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+    try:
+        logger.info(f"Saving regions to: {output_path}")
+
+        # Ensure output directory exists
+        output_dir = Path(output_path).parent
+        output_dir.mkdir(parents=True, exist_ok=True)
+        logger.info(f"Created directory: {output_dir}")
```
this isn't needed if snakemake has this output defined

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+        regions_gdf (geopandas.GeoDataFrame): Regions to save
+        output_path (str): Output file path
+    """
+    try:
+        logger.info(f"Saving regions to: {output_path}")
```
remove

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+            logger.warning("No regions contain powerplants! Keeping all regions.")
+            return regions_gdf
+            
+    except Exception as e:
+        logger.error(f"Error filtering regions with powerplants: {e}")
+        logger.warning("Continuing with unfiltered regions")
+        return regions_gdf
```
remove, rather let the workflow fail

yerbol-akhmetov replied:
Thanks for pointing, @FabianHofmann. I will remove the function altogether, because that function is no longer in use.

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+    """
+    logger.info(f"Loading powerplants from: {powerplants_path}")
+    
+    try:
```
remove, rather let the workflow fail with a clear error message

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+            f"Successfully created {len(regions_gdf)} regions from {len(country_shapes)} original shapes"
+        )
+        return regions_gdf
+    else:
+        logger.error("No regions were created!")
+        return country_shapes.copy()
```
remove, rather let the workflow fail with a clear error message

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+            f"Boundary lines converted. Length range: {boundary_lines.geometry.length.min():.0f} - {boundary_lines.geometry.length.max():.0f} meters"
+        )
+
+    if len(boundary_lines) == 0:
+        logger.warning("No valid boundary lines found!")
+        return country_shapes.copy()
```
replace with an error message that the function is expecting a non-empty bounary_lines dataframe

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+
+    if country_shapes.crs != target_crs:
+        country_shapes = country_shapes.to_crs(target_crs)
+        logger.info(
+            f"Country shapes converted. New total area: {country_shapes.geometry.area.sum() / 1000000:.0f} km²"
+        )
```
(personal taste) I would remove it as the logging shell is typically overcrowded in snakemake runs. you could instead use the debug mode 

```suggestion
        logger.debug(
            f"Country shapes converted. New total area: {country_shapes.geometry.area.sum() / 1000000:.0f} km²"
        )
```

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+        boundary_gdf = gpd.read_file(filepath)
+        logger.info(f"Loaded {len(boundary_gdf)} boundary features")
+        return boundary_gdf
+    except Exception as e:
+        logger.error(f"Error loading boundary lines: {e}")
+        raise
```
remove the try-except and rather let the workflow fail with a clear error message

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+        raise
+
+
+def ensure_same_crs(gdf1, gdf2):
+    """
```
```suggestion
def align_crs(gdf1, gdf2):
    """
```

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+    regions = []
+    region_id = 1
+
+    for idx, country in country_shapes.iterrows():
+        logger.info(f"Processing country/region {idx + 1}/{len(country_shapes)}")
+
```
this is quite a for-loop, is there any possibility to use geopandas built-in functions for this? In case boundary lines are line geometries you could do something leaner like

```
import geopandas as gpd # move to top 
from shapely.ops import split, unary_union # move to top


# merge all lines into one MultiLineString
all_lines = unary_union(gdf_lines.geometry)

# storage for results
split_geoms = []

for country in country_shapes.geometry:
    # split polygon by all lines
    try:
        result = split(country, all_lines)
        split_geoms.extend(result.geoms)  # add all parts
    except Exception:
        # if no intersection, keep original
        split_geoms.append(country)

# build new GeoDataFrame
regions = gpd.GeoDataFrame(geometry=split_geoms, crs=gdf_polys.crs)
```

yerbol-akhmetov replied:
Hi, @FabianHofmann, thank you for the comment. Probably it is possible. I think it would be better to point out this task and come back to it later, because there are currently other tasks. So we will not hold this PR.

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+    return regions_gdf
+
+
+def filter_regions_with_powerplants(regions_gdf, powerplants_path):
```
could you point out why this is needed?

FabianHofmann commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,421 @@
+        return country_shapes.copy()
+
+
+def clean_regions(regions_gdf, min_area_threshold=1000):
```
probably something more concrete would be good 
```suggestion
def drop_small_regions(regions_gdf, min_area_threshold=1000):
```

FabianHofmann reviewed (commented):
thanks @yerbol-akhmetov looks good! 

In general I would tend to 
* convert the more `logging.info` statements to `logging.debug` or remove them entirely in order not to bloat the output in the shell
* remove the try-except block where possible and rather let the workflow fail, I only marked some of them

I it hard for me to review the manual merge script. can we expect that to be long living, ie do we expect that the input to this might change and therefore the whole script. If so, I might be worth thinking about a more general approach

yerbol-akhmetov commented:
> thanks @yerbol-akhmetov looks good!
> 
> In general I would tend to
> 
> * convert the more `logging.info` statements to `logging.debug` or remove them entirely in order not to bloat the output in the shell
> * remove the try-except block where possible and rather let the workflow fail, I only marked some of them
> 
> I it hard for me to review the manual merge script. can we expect that to be long living, ie do we expect that the input to this might change and therefore the whole script. If so, I might be worth thinking about a more general approach

Hi, @FabianHofmann. Thank you for the comment. `manual_region_merger` script performs additional manual splitting of the regions (suggested by the expert), and merging regions. I have tried to do a more general automated way of region splitting and merging. But it did not worked out. Regions were not split well, and merging was not as expected. We have tried to find regions where powerplants exist. The resultant figure was accepted, but regions in between needed to be filled. I have spent few days with agentic mode and it could not provide satisfiable result. Finally, it was decided to just manually merge regions by providing grouping list.

In the future, they said that they may slightly alter regions. But I think it would be better to focus on that as problem arises, rather than spending time on it now to make it general. Of course, general approach is good, but we have already spent too much time on it.

yerbol-akhmetov commented on config/config.GB.yaml:
```diff
@@ -0,0 +1,119 @@
+  - 2030
+
+# docs in https://pypsa-eur.readthedocs.io/en/latest/configuration.html#countries
+countries:
+- BE
+- DK
+- FR
+- GB
+- IE
+- NL
+- 'NO'
```
Thank you for the comment, @brynpickering. The purpose of this config file is for the test, and it is not particular for this PR. I have included only neighboring countries. This test config can be used to extract transmission lines and buses for GB model (pre add_electricity). I will rename it as `config.GB.test.yaml` or so. It is temporary file. Of course, the scenario config will be different.

yerbol-akhmetov commented on config/config.GB.yaml:
```diff
@@ -0,0 +1,119 @@
+
+scenario:
+  clusters:
+  - 100
```
I have just set it 100 nodes, with focus weights set to make sure neighboring countries will have 1 node per interconnector. This config was for initial tests to get country shapes. But I will remove the config.

yerbol-akhmetov commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,307 @@
+# SPDX-FileCopyrightText:  Open Energy Transition gGmbH
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
```
Hi, @brynpickering. Not sure about this point. We always used to define Open Energy Transition as a copyright holder. In addition, we used to be more inclined on AGPL3 license. Also, we are not yest decided on license. Maybe we can discuss it internally.

yerbol-akhmetov commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,307 @@
+    """
+    # Convert to a projected CRS for accurate area calculations
+    # Use British National Grid (EPSG:27700) which is appropriate for UK
+    target_crs = "EPSG:27700"
```
Hi, @brynpickering. Thanks for the comment. I have put `target_crs` parameter to `config/config.common.yaml`. This will be configfile that are common for all scenarios.

yerbol-akhmetov commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,307 @@
+    if "snakemake" not in globals():
+        from scripts._helpers import mock_snakemake
+
+        snakemake = mock_snakemake("create_region_shapes", configfiles="config/config.GB.yaml")
```
Thanks for the comment, @brynpickering. I will remove the configfiles from here. In the `gb-model.smk` we will have common configs for scenarios. But suppose we would like to simulate different scenarios (set of reinforcements), we will probably have different names and folders for them. Currently, it is there for debugging purpose. I think it is better to remove it. When we need to debug based on certain scenario config, we will add it by ourselves locally.

yerbol-akhmetov commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,307 @@
+        logger.debug(
+            f"Successfully created {len(regions_gdf)} regions from {len(country_shapes)} original shapes"
+        )
+        return regions_gdf
```
Thank you, @brynpickering. I have checked the difference between two shapes. The area loss is 4.5km2. It is because we had an option to drop areas less than 1km2. In terms of percentage, it is 0.0019% loss. I am adding the part to calculate the difference and raise error, if difference is more than 0.1%.

yerbol-akhmetov commented on scripts/gb-model/create_region_shapes.py:
```diff
@@ -0,0 +1,307 @@
+        logger.debug(f"- Average area: {regions.geometry.area.mean() / 1000000:.1f} km²")
+
+    # Clean regions with appropriate threshold (1 km²)
+    min_area = 1000000  # 1 km² in square meters
```
@brynpickering, thank you, I have made it configurable.

yerbol-akhmetov commented on scripts/gb-model/manual_region_merger.py:
```diff
@@ -0,0 +1,504 @@
+    [74, 75, 76],                               # Join 74, 75, 76
+    [4, '5w']                                   # Join 4, 5w
+    # Keep 46n, 5e, and 8e as separate regions (no merge groups needed)
+]
```
Thanks, I have moved them to config.

yerbol-akhmetov commented on scripts/gb-model/manual_region_merger.py:
```diff
@@ -0,0 +1,466 @@
+                new_region['numeric_id'] = f"{region_num}e"
+
+                if 'area_km2' in new_region.index:
+                    new_area_km2 = geom.area / 1000000 if regions_gdf.crs != 'EPSG:4326' else geom.area * 111000 * 111000 / 1000000
```
I have implemented similar thing. But problems arose related to some `regions_gdf` not containing crs in some cases. I have also made a separate function of it to reuse it in all places. But I need to debug to see where is the problem.

yerbol-akhmetov commented on scripts/gb-model/manual_region_merger.py:
```diff
@@ -0,0 +1,466 @@
+                    new_region['numeric_id'] = f"{region_num}{suffixes[i]}"
+                    
+                    if 'area_km2' in new_region.index:
+                        new_area_km2 = geom.area / 1000000 if regions_gdf.crs != 'EPSG:4326' else geom.area * 111000 * 111000 / 1000000
```
Hi, @brynpickering. Thanks for the comment. I have applied your suggestion and debugged it. Now it works fine.

yerbol-akhmetov commented on config/config.GB.yaml:
```diff
@@ -0,0 +1,119 @@
+# SPDX-FileCopyrightText: Open Energy Transition gGmbH and contributors to PyPSA-Eur <https://github.com/pypsa/pypsa-eur>
+#
```
Thank you, @brynpickering. This is `config.GB.yaml` that I forgot to remove. So I will just remove it.

yerbol-akhmetov commented:
> Minor comment. Also, why is the GB config now deleted? I assume we want to keep a minimal GB config?

Hi, @brynpickering. Thanks for the comments. I have seen that `config.GB.yaml` raised too many questions. Therefore, I have removed it. The sole purpose of it was for you to test the PR as I have noted in the comments. We need to think about parameters of the future stable config. If it is needed I can bring `config.GB.yaml` back. However, I will strip everything, except a few lines.
